### PR TITLE
Fixed #17905 -- Restricted access to model pages in admindocs to users with permissions.

### PR DESCRIPTION
**Pull request by sai-ganesh-03**
#### Trac ticket number

ticket-17905

#### Branch description
Admin documentation lists only the models that users have access to

#### Checklist
- [x] This PR targets the `main` branch. 
- [ ] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

### CODE REVIEW

File: django/contrib/admindocs/views.py [hunk 1]
```diff
@@ -205,15 +210,67 @@ def get_context_data(self, **kwargs):
 class ModelIndexView(BaseAdminDocsView):
     template_name = "admin_doc/model_index.html"
 
+    def _user_has_model_perms(self, app_label, model_name, obj=None):
```
Comment (sai-ganesh-03) [inline django/contrib/admindocs/views.py:213]: Should I add this function as a method to [AnonymousUser](https://github.com/django/django/blob/661dfdd59809f4abd5077f7a2529735d07b98ba4/django/contrib/auth/models.py#L529). Just like the method [has_module_perm](https://github.com/django/django/blob/661dfdd59809f4abd5077f7a2529735d07b98ba4/django/contrib/auth/models.py#L619) ???

Comment (sarahboyce) [reply django/contrib/admindocs/views.py:213]: No
I think we want the logic to be as close as possible to `ModelAdmin` so I would do something like:
```diff
--- a/django/contrib/admindocs/views.py
+++ b/django/contrib/admindocs/views.py
@@ -1,5 +1,4 @@
 import inspect
-import re
 from importlib import import_module
 from inspect import cleandoc
 from pathlib import Path
@@ -14,6 +13,7 @@ from django.contrib.admindocs.utils import (
     replace_named_groups,
     replace_unnamed_groups,
 )
+from django.contrib.auth import get_permission_codename
 from django.core.exceptions import (
     ImproperlyConfigured,
     PermissionDenied,
@@ -206,39 +206,24 @@ class ViewDetailView(BaseAdminDocsView):
             }
         )
 
+def user_has_model_view_permission(user, opts):
+    """Based off ModelAdmin.has_view_permission."""
+    codename_view = get_permission_codename("view", opts)
+    codename_change = get_permission_codename("change", opts)
+    return user.has_perm(
+        "%s.%s" % (opts.app_label, codename_view)
+    ) or user.has_perm("%s.%s" % (opts.app_label, codename_change))
+
 
 class ModelIndexView(BaseAdminDocsView):
     template_name = "admin_doc/model_index.html"
 
-    def _user_has_model_perms(self, app_label, model_name, obj=None):
-        user_permissions = self.request.user.get_all_permissions()
-        # Regex pattern
-        pattern = r"(?P<appname>.+?)\.(?P<permission>.+?)_(?P<modelname>.+)"
-
-        # Set to hold app.model names
-        app_model_names_from_perms = set()
-
-        # Iterate over user permissions and match the pattern
-        for user_permission in user_permissions:
-            match = re.match(pattern, user_permission)
-            if match:
-                appname = match.group("appname")
-                modelname = match.group("modelname")
-                # Construct the app.model string and add it to the set
-                app_model_names_from_perms.add(f"{appname}.{modelname}")
-
-        app_model_name = f"{app_label}.{model_name}"
-        if app_model_name.lower() in app_model_names_from_perms:
-            return True
-
-        return False
-
     def get_context_data(self, **kwargs):
         m_list = []
         for m in apps.get_models():
             if self.request.user.has_module_perms(
                 m._meta.app_label
-            ) and self._user_has_model_perms(m._meta.app_label, m._meta.model_name):
+            ) and user_has_model_view_permission(self.request.user, m._meta):
                 m_list.append(m._meta)
         return super().get_context_data(**{**kwargs, "models": m_list})
 
@@ -246,31 +231,7 @@ class ModelIndexView(BaseAdminDocsView):
 class ModelDetailView(BaseAdminDocsView):
     template_name = "admin_doc/model_detail.html"
 
-    def _user_has_model_perms(self, app_label, model_name, obj=None):
-        user_permissions = self.request.user.get_all_permissions()
-        # Regex pattern
-        pattern = r"(?P<appname>.+?)\.(?P<permission>.+?)_(?P<modelname>.+)"
-
-        # Set to hold app.model names
-        app_model_names_from_perms = set()
-
-        # Iterate over user permissions and match the pattern
-        for user_permission in user_permissions:
-            match = re.match(pattern, user_permission)
-            if match:
-                appname = match.group("appname")
-                modelname = match.group("modelname")
-                # Construct the app.model string and add it to the set
-                app_model_names_from_perms.add(f"{appname}.{modelname}")
-        app_model_name = f"{app_label}.{model_name}"
-
-        if app_model_name.lower() in app_model_names_from_perms:
-            return True
-
-        return False
-
     def get_context_data(self, **kwargs):
-        app_label = self.kwargs["app_label"]
         model_name = self.kwargs["model_name"]
         # Get the model class.
         try:
@@ -284,13 +245,11 @@ class ModelDetailView(BaseAdminDocsView):
                 _("Model %(model_name)r not found in app %(app_label)r") % self.kwargs
             )
 
-        if not self._user_has_model_perms(app_label, model_name):
-            raise PermissionDenied(
-                f"{self.request.user} does not have permission to this model"
-            )
-
         opts = model._meta
 
+        if not user_has_model_view_permission(self.request.user, opts):
+            raise PermissionDenied
+
         title, body, metadata = utils.parse_docstring(model.__doc__)
         title = title and utils.parse_rst(title, "model", _("model:") + model_name)
```

**Review by sarahboyce — COMMENTED** [review]
Thank you! I have some suggestions :+1:

**Review by sarahboyce — APPROVED** [review]
Thank you @sai-ganesh-03 :star: this looks good to me

**Pull request merged**